bundles/model/org.eclipse.smarthome.model.thing/src/org/eclipse/smarthome/model/thing/internal/GenericThingProvider.xtend

```diff
@@ -275,11 +276,30 @@ class GenericThingProvider extends AbstractProvider<Thing> implements ThingProvi
         val List<Channel> channels = newArrayList
         modelChannels.forEach [
             if (addedChannelIds.add(id)) {
-                val kind = if (it.channelKind == null) "State" else it.channelKind                 
-                val parsedKind = ChannelKind.parse(kind)                
-                val channel = ChannelBuilder.create(new ChannelUID(thingTypeUID, thingUID, id), type)
+                var ChannelKind parsedKind = ChannelKind.STATE
+                
+                var ChannelTypeUID channelTypeUID
+                var String itemType
+                if (it.channelType != null) {
+                    channelTypeUID = new ChannelTypeUID(thingUID.bindingId, it.channelType)
+                    val resolvedChannelType = TypeResolver.resolve(channelTypeUID)
+                    if (resolvedChannelType != null) {
+                        itemType = resolvedChannelType.itemType
+                        parsedKind = resolvedChannelType.kind
+                    } else {
+                        logger.error("Channel type {} could not be resolved.",  channelTypeUID.asString)
+                    }
+                } else {
+                    itemType = it.type
+                    
+                    val kind = if (it.channelKind == null) "State" else it.channelKind                 
+                    parsedKind = ChannelKind.parse(kind)
+                }
+                
+                var channel = ChannelBuilder.create(new ChannelUID(thingUID, id), itemType)
                     .withKind(parsedKind)
                     .withConfiguration(createConfiguration)
+                    .withType(channelTypeUID)
```

Comment:
I don't know anything about Xtext, but what about channelTypeUID if `it.channelType == null` (the else branch). It seems to be undefined / null. Is the call to `withType` okay for a non-defined type?